File: 2208.将数组和减半的最少操作次数.py
#
# @lc app=leetcode.cn id=2208 lang=python3
#
# [2208] 将数组和减半的最少操作次数
#

# @lc code=start
from typing import List
import heapq
class Solution:
    def halveArray(self, nums: List[int]) -> int:
        target = sum(nums) / 2
        ret = 0
        for i in range(len(nums)): nums[i] *= -1
        heapq.heapify(nums)
        while target > 0:
            tmp = heapq.heappop(nums) / 2
            heapq.heappush(nums, tmp)
            target += tmp
            ret += 1
        return ret
# @lc code=end

print(Solution().halveArray(nums = [5,19,8,1]))
print(Solution().halveArray(nums = [3,8,20]))

File: 2208.将数组和减半的最少操作次数.cpp
/*
 * @lc app=leetcode.cn id=2208 lang=cpp
 *
 * [2208] 将数组和减半的最少操作次数
 */

// @lc code=start
#include<vector>
#include<queue>
#include<numeric>
using namespace std;
class Solution {
public:
    int halveArray(const vector<int>& nums) {
        priority_queue<double> pq(nums.begin(), nums.end());
        int ret = 0;
        double target = accumulate(nums.begin(), nums.end(), 0.) / 2;
        while(target > 0){
            auto tmp = pq.top() / 2; pq.pop();
            pq.push(tmp);
            target -= tmp;
            ++ret;
        }
        return ret;
    }
};
// @lc code=end

